### PR TITLE
Add tail of pihole.log to debug output

### PR DESCRIPTION
**Pull request by yubiuser**
- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
Adds the tail of `pihole.log` to the debug output. So far only head was used, but sometimes it if necessary to see if `dnsmasq` is still running.

**How does this PR accomplish the above?:**
Takes `obfuscated_pihole_log` function out of `analyze_pihole_log` function and call it once with head and once with tail of `pihole.log`

**What documentation changes (if any) are needed to support this PR?:**
None

### CODE REVIEW

**Comment by dschaper** [comment]
I think we used to include the tail but it was removed for some reason. I don't remember the reason offhand and it's possible we never included the tail and I'm just deluded.

The check failures look like functions were removed from `basic install.sh` and the test were not updated to know that there were functions removed.

**Comment by PromoFaux** [comment]
Yeah, I've not touched the tests yet on the dev-v6 branch, I will add to the project plan when I'm next at the computer!

**Comment by yubiuser** [comment]
> I think we used to include the tail but it was removed for some reason. I don't remember the reason offhand and it's possible we never included the tail and I'm just deluded.

This was the PR (https://github.com/pi-hole/pi-hole/pull/1565) adding the current form of the log analysis. Although the commit (https://github.com/pi-hole/pi-hole/pull/1565/commits/fc0440546f95db37c3944ea40299f2e83235fe3e) was named "add functions to parse head and tails of gravity.list and pihole.log" `tail` of `pihole.log` was not included. Looking at the structure of this commit, it was likely only forgotten while copy&pasting...

**Comment by dschaper** [comment]
Okay, starting to recall what happened. We previously tailed `pihole.log` live and asked users to visit problem sites so that the log would capture the query and response. See https://github.com/pi-hole/pi-hole/pull/1565/files#diff-8e13fa9927aabae5439f4ae9d8736f56ab9220296dd931a69cab1e97830a63e3L405 for that. It required users to perform actions and didn't work with the web based debug process.

**Comment by MichaIng** [comment]
Keep in mind that this file logging can be disabled via `pihole -l off`

It might make sense to check first of it's enabled before adding the tail.

**Comment by yubiuser** [comment]
Good idea. However, even if `log-queries` is off, `dnsmasq` will log a few lines on start up which might be helpful for debugging. I just tried a debug run after setting `pihole -l off` and the file contained

```
-rw-r--r-- 1 pihole pihole 1905 Jun  3 13:07 /var/log/pihole.log
   -----head of pihole.log------
   Jun  3 13:07:17 dnsmasq[1254]: exiting on receipt of SIGTERM
   Jun  3 13:07:24 dnsmasq[30854]: started, version pi-hole-2.85 cachesize 10000
   Jun  3 13:07:24 dnsmasq[30854]: DNS service limited to local subnets
   Jun  3 13:07:24 dnsmasq[30854]: compile time options: IPv6 GNU-getopt no-DBus no-UBus no-i18n IDN DHCP DHCPv6 Lua TFTP no-conntrack ipset auth cryptohash DNSSEC loop-detect inotify dumpfile
   Jun  3 13:07:24 dnsmasq[30854]: using only locally-known addresses for domain use-application-dns.net
   Jun  3 13:07:24 dnsmasq[30854]: using nameserver 10.0.1.1#53 for domain fritz.box 
   Jun  3 13:07:24 dnsmasq[30854]: using nameserver 10.0.1.1#53 for domain 1.0.10.in-addr.arpa 
   Jun  3 13:07:24 dnsmasq[30854]: using nameserver 127.0.0.1#5335
   Jun  3 13:07:24 dnsmasq[30854]: read /etc/hosts - 7 addresses
   Jun  3 13:07:24 dnsmasq[30854]: read /etc/pihole/custom.list - 13 addresses
   Jun  3 13:07:24 dnsmasq[30854]: read /etc/pihole/local.list - 2 addresses
   Jun  3 13:07:24 dnsmasq[30854]: exiting on receipt of SIGTERM
   Jun  3 13:07:30 dnsmasq[30901]: started, version pi-hole-2.85 cachesize 10000
   Jun  3 13:07:30 dnsmasq[30901]: DNS service limited to local subnets
   Jun  3 13:07:30 dnsmasq[30901]: compile time options: IPv6 GNU-getopt no-DBus no-UBus no-i18n IDN DHCP DHCPv6 Lua TFTP no-conntrack ipset auth cryptohash DNSSEC loop-detect inotify dumpfile
   Jun  3 13:07:30 dnsmasq[30901]: using only locally-known addresses for domain use-application-dns.net
   Jun  3 13:07:30 dnsmasq[30901]: using nameserver 10.0.1.1#53 for domain fritz.box 
   Jun  3 13:07:30 dnsmasq[30901]: using nameserver 10.0.1.1#53 for domain 1.0.10.in-addr.arpa 
   Jun  3 13:07:30 dnsmasq[30901]: using nameserver 127.0.0.1#5335
   Jun  3 13:07:30 dnsmasq[30901]: read /etc/hosts - 7 addresses
```

**Comment by MichaIng** [comment]
Ah okay, then it still makes sense. The check could still serve as hint for the support team to not wonder where the query logs are, but not that important then.

**Comment by yubiuser** [comment]
I added the check, it will print an info about the disabled logging now.

```
*** [ DIAGNOSING ]: Pi-hole log
[i] Query logging is disabled

-rw-r--r-- 1 pihole pihole 953 Jun  4 21:12 /var/log/pihole.log
   -----head of pihole.log------
   Jun  4 21:12:44 dnsmasq[30901]: exiting on receipt of SIGTERM
   Jun  4 21:12:51 dnsmasq[23324]: started, version pi-hole-2.85 cachesize 10000
   Jun  4 21:12:51 dnsmasq[23324]: DNS service limited to local subnets
   Jun  4 21:12:51 dnsmasq[23324]: compile time options: IPv6 GNU-getopt no-DBus no-UBus no-i18n IDN DHCP DHCPv6 Lua TFTP no-conntrack ipset auth cryptohash DNSSEC loop-detect inotify dumpfile
   Jun  4 21:12:51 dnsmasq[23324]: using only locally-known addresses for domain use-application-dns.net
   Jun  4 21:12:51 dnsmasq[23324]: using nameserver 10.0.1.1#53 for domain fritz.box 
   Jun  4 21:12:51 dnsmasq[23324]: using nameserver 10.0.1.1#53 for domain 1.0.10.in-addr.arpa 
   Jun  4 21:12:51 dnsmasq[23324]: using nameserver 127.0.0.1#5335
   Jun  4 21:12:51 dnsmasq[23324]: read /etc/hosts - 7 addresses
   Jun  4 21:12:51 dnsmasq[23324]: read /etc/pihole/custom.list - 13 addresses
   Jun  4 21:12:51 dnsmasq[23324]: read /etc/pihole/local.list - 2 addresses

   -----tail of pihole.log------
   Jun  4 21:12:44 dnsmasq[30901]: exiting on receipt of SIGTERM
   Jun  4 21:12:51 dnsmasq[23324]: started, version pi-hole-2.85 cachesize 10000
   Jun  4 21:12:51 dnsmasq[23324]: DNS service limited to local subnets
   Jun  4 21:12:51 dnsmasq[23324]: compile time options: IPv6 GNU-getopt no-DBus no-UBus no-i18n IDN DHCP DHCPv6 Lua TFTP no-conntrack ipset auth cryptohash DNSSEC loop-detect inotify dumpfile
   Jun  4 21:12:51 dnsmasq[23324]: using only locally-known addresses for domain use-application-dns.net
   Jun  4 21:12:51 dnsmasq[23324]: using nameserver 10.0.1.1#53 for domain fritz.box 
   Jun  4 21:12:51 dnsmasq[23324]: using nameserver 10.0.1.1#53 for domain 1.0.10.in-addr.arpa 
   Jun  4 21:12:51 dnsmasq[23324]: using nameserver 127.0.0.1#5335
   Jun  4 21:12:51 dnsmasq[23324]: read /etc/hosts - 7 addresses
   Jun  4 21:12:51 dnsmasq[23324]: read /etc/pihole/custom.list - 13 addresses
   Jun  4 21:12:51 dnsmasq[23324]: read /etc/pihole/local.list - 2 addresses
```

**Comment by yubiuser** [comment]
Should I fix the stickler complain? I copied it from

https://github.com/pi-hole/pi-hole/blob/4736e03108763cc2d5659f48d8a1e8a64d9b2608/pihole#L358

**Comment by dschaper** [comment]
> Should I fix the stickler complain?

Yeah, go ahead and fix that. 

I'm not sure if this needs to be to development-v6 or just development. I don't want to push off improvements and fixes just to hit a specific version number.

**Comment by PromoFaux** [comment]
> I'm not sure if this needs to be to development-v6 or just development.

We are an an interesting crossroad here. I think it is probably more likely that a point release will be next out than v6 (though never say never) so it makes more sense to merge to `development` for the time being. I'll just try and keep on top of rebasing `development-v6`

**Comment by dschaper** [comment]
I wonder if there's an Action we could set up that would either auto rebase `v6` when `development` merges or auto opens a PR to merge `development` in to `v6`?

**Comment by yubiuser** [comment]
So I should rebase on `development`?

**Comment by MichaIng** [comment]
> I wonder if there's an Action we could set up that would either auto rebase `v6` when `development` merges or auto opens a PR to merge `development` in to `v6`?

There are at least actions for backporting PRs on merge, i.e. the commits from a merged PR will be automatically added to a new PR against another branch, usually based on labels: https://github.com/marketplace?query=backport
With the actions source code/methods, it should be easy to have a new PR with `development` branch as source created, but since not every commit might be wanted in `v6`, doing this per-PR is even better, I guess?

**Comment by dschaper** [comment]
I'm not sure why the tests are not firing. I think there may be something with just changing the base branches? But I'm not sure.

**Comment by dschaper** [comment]
> There are at least actions for backporting PRs on merge, i.e. the commits from a merged PR will be automatically added to a new PR against another branch,

That may be helpful in other places as well. I'm not sure how Promo wants to handle syncing the branches so I'll trust it to him to use a method that helps. If there's any way I can provide assistance then I will.

And Micha, thank you for contributing to Pi-hole. I know I've been mean and rude to you in the past and I've been seeing a lot of danger signs of burnout in myself which ends up in me being antagonistic and pushing back against the very thing that would help with the workload. Donations have been dropping for Pi-hole lately and that's a very nervous thing for me, that's not an excuse for my actions and behavior. Pi-hole was starting to not be fun for me anymore and I'm trying to get back to where it was fun and creative and open.

**Comment by MichaIng** [comment]
> And Micha, thank you for contributing to Pi-hole. I know I've been mean and rude to you in the past and I've been seeing a lot of danger signs of burnout in myself which ends up in me being antagonistic and pushing back against the very thing that would help with the workload. Donations have been dropping for Pi-hole lately and that's a very nervous thing for me, that's not an excuse for my actions and behavior. Pi-hole was starting to not be fun for me anymore and I'm trying to get back to where it was fun and creative and open.

Many thanks for your words, I highly appreciate it 👍. And I know I've also not always reacted fair and objective, interpreted reasonable requests in a negative way. Let's start fresh and productive. And you know, I'm bullhead enough to keep contributing to projects I find valuable, interesting and where I think I am able to help, even against headwind 🚀.

Best wishes and hopes for a successful future and development for the great project Pi-hole!

**Comment by dschaper** [comment]
>  And I know I've also not always reacted fair and objective, interpreted reasonable requests in a negative way.

That's a very fair assessment.

**Comment by PromoFaux** [comment]
Closing and reopening makes the tests go brrrrrrrrrrrrrr

File: advanced/Scripts/piholeDebug.sh [hunk 1]
```diff
@@ -1281,56 +1281,74 @@ analyze_gravity_list() {
...
+      # correctly, there will be an entry for "bad address at line n"
+      # So we can check for that here and highlight it in red so the user can see it easily
+      error_to_check_for=$(echo "${line}" | grep 'bad address at')
+      # Some users may not want to have the domains they visit sent to us
+      # To that end, we check for lines in the log that would contain a domain name
+      line_to_obfuscate=$(echo "${line}" | grep ': query\|: forwarded\|: reply')
+      # If the variable contains a value, it found an error in the log
+      if [[ -n ${error_to_check_for} ]]; then
+          # So we can print it in red to make it visible to the user
+          log_write "   ${CROSS} ${COL_RED}${head_line}${COL_NC} (${FAQ_BAD_ADDRESS})"
```
Comment (PromoFaux) [inline advanced/Scripts/piholeDebug.sh:1302]: Think I've found a bug...

image

image

But it's also in dev, so you didn't introduce it :)

**Review by PromoFaux — APPROVED** [review]

File: advanced/Scripts/piholeDebug.sh (hunk 1 above)
Comment (PromoFaux) [reply advanced/Scripts/piholeDebug.sh:1302]: https://github.com/pi-hole/pi-hole/pull/4216

**Pull request merged**